src/threading/cthread.hpp

```diff
@@ -0,0 +1,303 @@
+    // for wait
+    void run_tasks_while(task_group*);
+    // loop forever for secondary threads
+    // until quit is set
+    void run_tasks_loop();
+
+    // Create nthreads-1 new c std threads
+    // must be > 0
+    // singled only created in static get_global_task_pool()
+    task_pool(std::size_t nthreads);
+    
+    // task_pool is a singleton 
+    task_pool(const task_pool&) = delete;
+    task_pool& operator=(const task_pool&) = delete;
+
+    // set quit and wait for secondary threads to end
+    ~task_pool();
+  
+public:
+    // Like tbb calls: run queues a task,
+    // wait waits for all tasks in the group to be done
+    void run(const task&);
+    void run(task&&);
+    void wait(task_group*);
+  
+    // includes master thread
+    int get_num_threads() {
+        return _threads.size() + 1;
+    }
+
+    // get a stable integer for the current thread that
+    // is 0..nthreads
+    std::size_t get_current_thread() {
+        return _thread_ids[std::this_thread::get_id()];
+    }
+
+    // singleton constructor - needed to order construction
+    // with other singletons (profiler)
+    static task_pool& get_global_task_pool();
+};
+} //impl
+
+///////////////////////////////////////////////////////////////////////
+// types
+///////////////////////////////////////////////////////////////////////
+template <typename T>
+class enumerable_thread_specific {
+    impl::task_pool& global_task_pool;
+
+    using storage_class = std::vector<T>;
+    storage_class data;
+  
+public :
+    using iterator = typename storage_class::iterator;
+    using const_iterator = typename storage_class::const_iterator;
+
+    enumerable_thread_specific():
+        global_task_pool{impl::task_pool::get_global_task_pool()},
+        data{std::vector<T>(global_task_pool.get_num_threads())}
+    {}
+
+    enumerable_thread_specific(const T& init):
+        global_task_pool{impl::task_pool::get_global_task_pool()},
+        data{std::vector<T>(global_task_pool.get_num_threads(), init)}
+    {}
+
+    T& local() {
+      return data[global_task_pool.get_current_thread()];
+    }
+    const T& local() const {
+      return data[global_task_pool.get_current_thread()];
+    }
+
+    auto size() -> decltype(data.size()) const { return data.size(); }
+
+    iterator begin() { return data.begin(); }
+    iterator end()   { return data.end(); }
+
+    const_iterator begin() const { return data.begin(); }
+    const_iterator end()   const { return data.end(); }
+
+    const_iterator cbegin() const { return data.cbegin(); }
+    const_iterator cend()   const { return data.cend(); }
+};
+
+template <typename T>
+class parallel_vector {
+    using value_type = T;
+    std::vector<value_type> data_;
+
+private:
+    // lock the parallel_vector to update
+    impl::mutex mutex;
+
+    // call a function of type X f() in a lock
+    template<typename F>
+    auto critical(F f) -> decltype(f()) {
+        impl::lock lock{mutex};
+        return f();
+    }
+
```

Comment:
It could be a bit more general, but it's only used in one place, and lambdas anyway are an easy way to get around the void parameter.